src/app/app.component.ts

```diff
@@ -1,276 +1,338 @@
+  private readonly mailboxPrefix = '_mailbox';
+
+  // Cached mailbox and file data
+  mailboxes: MailboxView[] = [];
+  activeMailbox: MailboxView = null;
+  selectedFile: File;
+
+  ZaxMessageKind = ZaxMessageKind;
 
   // UI defaults
+  UIAction = UIAction;
+  UIFlags = {
+    [UIAction.keyAdded]: false,
+    [UIAction.mailboxCreated]: false,
+    [UIAction.messageSent]: false,
+    [UIAction.fileSent]: false,
+    [UIAction.messagesLoading]: false,
+    [UIAction.counterLoading]: false
+  };
   relayURL: string;
   editingURL: string;
+  isEditing = false;
   newMailboxSubscreen = 'new';
   viewMailboxSubscreen = 'inbox';
-  quantity = 5;
-
-  constructor(private http: HttpClient) {
-    this.initGlow();
-
-  }
-
-  ngOnInit() {
-    this.setDefaultRelay();
-    this.initRelay(this.relayURL);
-    this.initMailboxes();
-  }
 
   // -------------------------
   // Initialization
   // -------------------------
 
-  private initGlow() {
-    this.glow = (window as any).glow;
-    this.glow.CryptoStorage.startStorageSystem(new this.glow.SimpleStorageDriver());
-
-    this.mailbox = this.glow.MailBox;
-
-    this.glow.setAjaxImpl((url: string, data: string) => {
-      const request = this.http.post(url, data, {
-        headers: {
-          'Accept': 'text/plain',
-          'Content-Type': 'text/plain'
-        },
-        responseType: 'text'
-      });
-
-      return request.toPromise();
-    });
+  async ngOnInit(): Promise<void> {
+    NaCl.setInstance();
+    this.setDefaultRelay();
+    await this.initMailboxes();
   }
 
   private setDefaultRelay() {
-    if (window.location.origin.indexOf('github.io') > -1) {
-      // Use test server by default when running on http://vault12.github.io/zax-dashboard/
+    if (window.location.origin.indexOf('github.io') > -1 || window.location.origin.indexOf('localhost') > -1) {
+      // Use test server by default when running on http://vault12.github.io/zax-dashboard/ or locally
       this.relayURL = 'https://z.vault12.com';
     } else {
       // Use current location otherwise
       // NOTE: Take care not to mix up ports when both are running locally
-      this.relayURL = 'https://z2.vault12.com'; //window.location.origin;
+      this.relayURL = window.location.origin;
     }
     this.editingURL = this.relayURL;
   }
 
-  private initRelay(url: string) {
-    this.relay = new this.glow.Relay(url);
-  }
-
-  async initMailboxes() {
+  private async initMailboxes(): Promise<void> {
     // add all mailboxes stored in localStorage
-    var mbx_count = 0
-    var localKeys = Object.keys(localStorage)
+    const localKeys = Object.keys(localStorage).filter(key => key.startsWith(this.mailboxPrefix));
     for (const key of localKeys) {
-      if (key.indexOf(this.mailboxPrefix) === 0) {
-        mbx_count++
-        await this.generateMailbox(localStorage.getItem(key));
-      }
+      await this.generateMailbox(localStorage.getItem(key));
```

Comment:
so every time page reloads we create new mailboxes but with same id. is it ok?

Reply:
Right, IDs and Glow's CryptoStorage are stored in browser's localStorage, and the app simply reads serialized data and restores them when the page loads. Do you have better ideas?